Guards typecheck

reilem opened the pull request.

reilem committed:
New FcGuards, FcGuarded types + pretty printing & contains free vars

reilem committed:
Implement substitutions for new Fc Types

reilem committed:
Fix renamer warning: name shadowing

reilem committed:
Fix new FcTypes phases

reilem committed:
Implement freshenLclBndrs for new types

reilem committed:
Fix elabClsDecl and elabMethodSig problem

reilem committed:
Pattern guard initial typechecking

reilem committed:
Add desugaring holes

reilem committed:
Small fix nested-up-guards test

reilem committed:
Add renamer and typecheck failures + added options to guardTest script

reilem committed:
Fix free type variable gathering for pattern guards

gkaracha commented on Backend/FcTypes.hs:
```diff
@@ -261,7 +268,20 @@ instance ContainsFreeTyVars (FcTerm a) FcTyVar where
+  ftyvsOf (FcConPat   _dc _xs) = []
+  ftyvsOf (FcConPatNs _dc _ps) = []
+  ftyvsOf (FcVarPat   _x)      = []
+  ftyvsOf (FcOrPat    _p1 _p2) = []
```
The "truly" generic solution is the following (essentially you do recurse even if all of them end up returning [] at the end):
```suggestion
instance ContainsFreeTyVars (FcPat a) FcTyVar where
  ftyvsOf (FcConPat   _dc _xs) = []
  ftyvsOf (FcConPatNs _dc ps) = concatMap ftyvsOf ps
  ftyvsOf (FcVarPat   _x)      = []
  ftyvsOf (FcOrPat    p1 p2) = ftyvsOf p1 ++ ftyvsOf p2
```

gkaracha commented on Backend/FcTypes.hs:
```diff
@@ -261,7 +268,20 @@ instance ContainsFreeTyVars (FcTerm a) FcTyVar where
 instance ContainsFreeTyVars (FcAlt a) FcTyVar where
-  ftyvsOf (FcAlt _pat tm) = ftyvsOf tm
+  ftyvsOf (FcAltFc _pat tm)  = ftyvsOf tm
+  ftyvsOf (FcAltTc _pat gRs) = concatMap ftyvsOf gRs
```
The "truly" generic solution is the following (essentially you do recurse even if all of them end up returning [] at the end):
```suggestion
  ftyvsOf (FcAltFc pat tm)  = ftyvsOf pat ++ ftyvsOf tm
  ftyvsOf (FcAltTc pat gRs) = ftyvsOf pat ++ concatMap ftyvsOf gRs
```

gkaracha commented on Backend/FcTypes.hs:
```diff
@@ -352,9 +372,20 @@ instance PrettyPrint (FcPat a) where
 instance PrettyPrint (FcAlt a) where
-  ppr (FcAlt p tm) = ppr p <+> arrow <+> ppr tm
+  ppr (FcAltFc p tm)  = ppr p <+> arrow <+> ppr tm
+  ppr (FcAltTc p gRs) = ppr p <+> (vcat $ map ppr gRs)
```
Alternatively, you can make the guarded equations "hang" (like how you'd do in your own code):
```suggestion
  ppr (FcAltTc p gRs) = hang (ppr p) 2 (vcat $ map ppr gRs)  -- I usually use 2 spaces
```

gkaracha replied:
Also, I think you missed the bar `|` at the start of each guarded equation :wink:

gkaracha commented on Backend/FcTypes.hs:
```diff
@@ -179,6 +179,12 @@ data FcTerm (a :: Phase) where
 -- you do it means that something is probably wrong (the only setting where you
 -- need stuff like this is for optimizations).
 
+data FcGuard (a :: Phase) where
+  FcPatGuard :: FcPat 'Tc -> FcTerm 'Tc -> FcGuard 'Tc
+
+data FcGuarded (a :: Phase) where
+  FcGuarded :: [FcGuard 'Tc] -> FcTerm 'Tc -> FcGuarded 'Tc
```
Note that these two don't have to be GADTs (actually they don't need an index at all), since they don't have---and will never have---a counterpart in the target language. You can leave them like this for uniformity, I just wanted to point this out :slightly_smiling_face:

reilem replied:
Indeed, I thought so as well. But I got a bunch of errors in places when I didn't give them an index.

reilem replied:
More specifically:
```
/Users/Reinert/Documents/Coding/Haskell/Thesis/KHC/Utils/Substitution.hs:154:10: error:
    • Illegal instance declaration for
        ‘SubstVar FcTmVar (FcTerm a) FcGuarded’
        The liberal coverage condition fails in class ‘SubstVar’
          for functional dependency: ‘v x -> t’
        Reason: lhs types ‘FcTmVar’, ‘FcGuarded’
          do not jointly determine rhs type ‘FcTerm a’
        Un-determined variable: a
    • In the instance declaration for
        ‘SubstVar FcTmVar (FcTerm a) FcGuarded’
    |
154 | instance SubstVar FcTmVar (FcTerm a) FcGuarded where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

/Users/Reinert/Documents/Coding/Haskell/Thesis/KHC/Utils/Substitution.hs:157:10: error:
    • Illegal instance declaration for
        ‘SubstVar FcTmVar (FcTerm a) FcGuard’
        The liberal coverage condition fails in class ‘SubstVar’
          for functional dependency: ‘v x -> t’
        Reason: lhs types ‘FcTmVar’, ‘FcGuard’
          do not jointly determine rhs type ‘FcTerm a’
        Un-determined variable: a
    • In the instance declaration for
        ‘SubstVar FcTmVar (FcTerm a) FcGuard’
    |
157 | instance SubstVar FcTmVar (FcTerm a) FcGuard where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Failed, 15 modules loaded.
```

gkaracha commented on Utils/PrettyPrint.hs:
```diff
@@ -163,10 +163,11 @@ rbrack = liftDoc P.rbrack
+underscore, dcolon, arrow, larrow, darrow, dot, backslash :: Doc
 underscore = liftDoc (P.char '_')
 dcolon     = colorDoc yellow $ liftDoc (P.text "::")
 arrow      = colorDoc yellow $ liftDoc (P.text "->")
```
to be consistent, now I would rename the plain `arrow` to `rarrow`.

gkaracha commented on Frontend/HsTypeChecker.hs:
```diff
@@ -504,29 +505,44 @@ elabTmCase scr alts = do
+
+elabHsGuard :: RnGuard -> GenM (RnBinds, FcGuard 'Tc)
+elabHsGuard (HsPatGuard p tm) = do
+  (tm_ty, fc_tm) <- elabTerm tm
+  (binds, fc_p)  <- elabHsPat tm_ty p []
```
Here you also need to make sure that the type of the pattern matches the type of the expression.

reilem replied:
I think elabHsPat already does this

gkaracha replied:
You are correct.

gkaracha commented on Utils/Substitution.hs:
```diff
@@ -105,8 +106,12 @@ instance SubstVar FcTyVar FcType (FcTerm a) where
 -- | Substitute a type variable for a type in a case alternative
 instance SubstVar FcTyVar FcType (FcAlt a) where
-  substVar a ty (FcAlt p tm) = FcAlt p (substVar a ty tm)
+  substVar a ty (FcAltFc p tm)  = FcAltFc p (substVar a ty tm)
```
For consistency (the extensibility that we talked about):
```suggestion
  substVar a ty (FcAltFc p tm)  = FcAltFc (substVar a ty p) (substVar a ty tm)
```

gkaracha commented on Utils/Substitution.hs:
```diff
@@ -105,8 +106,12 @@ instance SubstVar FcTyVar FcType (FcTerm a) where
+  substVar a ty (FcAltTc p gRs) = FcAltTc p (map (substVar a ty) gRs)
+
+instance SubstVar FcTyVar FcType (FcGuarded a) where
+  substVar a ty (FcGuarded gs tm) = FcGuarded gs (substVar a ty tm)
```
For consistency (the extensibility that we talked about):
```suggestion
  substVar a ty (FcGuarded gs tm) = FcGuarded (map (substVar a ty) gs) (substVar a ty tm)
```

gkaracha commented on Utils/Substitution.hs:
```diff
@@ -389,14 +401,35 @@ instance FreshenLclBndrs (FcTerm a) where
+    return (FcAltTc p' gRs')
+
+instance FreshenLclBndrs (FcGuarded a) where
+  freshenLclBndrs (FcGuarded []     tm) = return (FcGuarded [] tm)
```
You forgot to freshen the binders in the expression here:
```suggestion
  freshenLclBndrs (FcGuarded []     tm) = return (FcGuarded [] (freshenLclBndrs tm))
```

gkaracha commented on Utils/Substitution.hs:
```diff
@@ -389,14 +401,35 @@ instance FreshenLclBndrs (FcTerm a) where
+
+freshenGuardLclBndrs :: MonadUnique m => FcGuard a -> m (FcGuard a, BndrSubsts)
+freshenGuardLclBndrs (FcPatGuard p tm) = do
+  (p', substs) <- freshenPatLclBndrs p
```
Again, you forgot to freshen `tm` here.

gkaracha commented on Utils/Substitution.hs:
```diff
@@ -389,14 +401,35 @@ instance FreshenLclBndrs (FcTerm a) where
+instance FreshenLclBndrs (FcGuarded a) where
+  freshenLclBndrs (FcGuarded []     tm) = return (FcGuarded [] tm)
+  freshenLclBndrs (FcGuarded (g:gs) tm) = do
+    (g', substs)        <- freshenGuardLclBndrs g
+    (FcGuarded gs' tm') <- freshenLclBndrs (FcGuarded (map (doSubsts substs) gs) (doSubsts substs tm))
```
You can remove the parentheses here:
```suggestion
    (g', substs)      <- freshenGuardLclBndrs g
    FcGuarded gs' tm' <- freshenLclBndrs (FcGuarded (map (doSubsts substs) gs) (doSubsts substs tm))
```

gkaracha reviewed (changes requested):
I hope I didn't miss anything!

reilem committed:
More generic ftyvsOf + rarrow, bar and hang pretty printing

reilem committed:
Fix George's suggestions

Fix freshening issues
Make FcType substitution more generic
https://github.com/reilem/KHC/pull/24#pullrequestreview-386612147

reilem committed:
Fix pretty printing
